test/metatx/ERC2771Forwarder.t.sol

```diff
@@ -0,0 +1,158 @@
+        return
+            _hashTypedDataV4(
+                keccak256(
+                    abi.encode(
+                        _FORWARD_REQUEST_TYPEHASH,
+                        request.from,
+                        request.to,
+                        request.value,
+                        request.gas,
+                        request.nonce,
+                        request.deadline,
+                        keccak256(request.data)
+                    )
+                )
+            );
+    }
+}
+
+contract ERC2771ForwarderTest is Test {
+    ERC2771ForwarderMock internal _erc2771Forwarder;
+    CallReceiverMock internal _receiver;
+
+    uint256 internal _signerPrivateKey;
+    uint256 internal _relayerPrivateKey;
+
+    address internal _signer;
+    address internal _relayer;
+
+    uint256 internal constant _MAX_ETHER = 10_000_000; // To avoid overflow
+
+    function setUp() public {
+        _erc2771Forwarder = new ERC2771ForwarderMock("ERC2771Forwarder");
+        _receiver = new CallReceiverMock();
+
+        _signerPrivateKey = 0xA11CE;
+        _relayerPrivateKey = 0xB0B;
+
+        _signer = vm.addr(_signerPrivateKey);
+        _relayer = vm.addr(_relayerPrivateKey);
+    }
+
+    function _forgeRequestData(
+        uint256 value,
+        uint256 nonce,
+        uint48 deadline,
+        bytes memory data
+    ) private view returns (ERC2771Forwarder.ForwardRequestData memory _requestData) {
+        ForwardRequest memory request = ForwardRequest({
+            from: _signer,
+            to: address(_receiver),
+            value: value,
+            gas: 30000,
+            nonce: nonce,
+            deadline: deadline,
+            data: data
+        });
+
+        bytes32 digest = _erc2771Forwarder.structHash(request);
+        (uint8 v, bytes32 r, bytes32 s) = vm.sign(_signerPrivateKey, digest);
+        bytes memory signature = abi.encodePacked(r, s, v);
+
+        _requestData = ERC2771Forwarder.ForwardRequestData({
+            from: request.from,
+            to: request.to,
+            value: request.value,
+            gas: request.gas,
+            deadline: request.deadline,
+            data: request.data,
+            signature: signature
+        });
+    }
+
+    function testExecuteAvoidsETHStuck(uint256 initialBalance, uint256 value) public {
+        vm.assume(initialBalance < _MAX_ETHER);
+        vm.assume(value < _MAX_ETHER);
+
+        vm.deal(address(_erc2771Forwarder), initialBalance);
+
+        uint256 nonce = _erc2771Forwarder.nonces(_signer);
+
+        vm.deal(address(this), value);
+        _erc2771Forwarder.execute{value: value}(
+            _forgeRequestData({
+                value: value,
+                nonce: nonce,
+                deadline: uint48(block.timestamp + 1),
+                data: abi.encodeCall(CallReceiverMock.mockFunction, ())
+            })
+        );
+
+        assertEq(address(_erc2771Forwarder).balance, initialBalance);
+    }
```

Comment:
This test is fuzzing the initialBalance and the value, but not the "paths" in the `execute` function. It only runs a function that we know is ok, with a deadline we know is ok, and a nonce we know is correct, with a fixed amount of gas

Ideally, we would want to consider many data, to check that whatever the outcome of the subcall, balances are preserved.

Maybe you could have a uint8 argument that is used to select one of "many" function. `mockFunction` will only be one of them. It should include: non payable functions, function that revert, function that panic, ...

Reply:
or just do 50/50 between `mockFunction` and `mockFunctionRevertsNoReason`

Reply:
You're right, ideally, we would fuzz multiple functions. However, it becomes difficult to conditionally expect a revert in the test, so I did the 50/50 between `mockFunction` and `mockFunctionRevertNoReason` you proposed to cover both paths.

The `deadline`, `nonce` and `gas` are fixed because the contract is about making sure there's no value left in the contract and I don't think there's a way the `deadline` or `nonce` might affect that. However, `gas` may affect it because of the `_checkForwardedGas` but it's already proven in the comments.

I think the suggestion you proposed for this test is enough for its purpose (not leaving ETH stuck). I also tried to fuzz the `gas` very early in #4346 but it becomes messier because we have to conditionally expect reverts as well, and it's difficult to predict. Would appreciate if you have any idea how it may work in case you think it's a valuable addition